File: swift/EgyNames/Sources/EgyNames/Splitter.swift
import Foundation

public final class Splitter: @unchecked Sendable {
    private static let BASE_SEGMENT_COST = 1.0
    private static let UNKNOWN_PENALTY = 8.0
    private static let LENGTH_BONUS_PER_CHAR = -0.05

    private static func getFreqBonus(_ fc: FrequencyClass) -> Double {
        switch fc {
        case .common: return -0.6
        case .normal: return -0.2
        case .rare: return 0.0
        }
    }

    public static func dpSegment(_ text: String, customPath: String? = nil) -> [String] {
        LookupIndices.ensureBuilt(customPath: customPath)
        let chars = Array(text)
        let n = chars.count
        guard n > 0 else { return [text] }

        let arNorm = LookupIndices.getArNormForms(customPath: customPath)

        var dpCost = Array(repeating: Double.infinity, count: n + 1)
        var dpPrev = Array(repeating: -1, count: n + 1)

        dpCost[0] = 0.0
        dpPrev[0] = 0

        for i in 1...n {
            let startJ = (i > 45) ? (i - 45) : 0
            for j in startJ..<i {
                if dpCost[j] == Double.infinity { continue }

                let substr = String(chars[j..<i])
                if substr.count < 2 && j > 0 { continue }

                let norm = LookupIndices.normalizeAr(substr)
                if let entry = arNorm[norm] {
                    let cost = dpCost[j] + BASE_SEGMENT_COST + getFreqBonus(entry.frequency) + LENGTH_BONUS_PER_CHAR * Double(substr.count)
                    if cost < dpCost[i] {
                        dpCost[i] = cost
                        dpPrev[i] = j
                    }
                } else {
                    let cost = dpCost[j] + UNKNOWN_PENALTY + Double(substr.count)
                    if cost < dpCost[i] {
                        dpCost[i] = cost
                        dpPrev[i] = j
                    }
                }
            }
        }

        if dpCost[n] == Double.infinity {
            return [text]
        }

        var segments: [String] = []
        var pos = n
        while pos > 0 {
            let prev = dpPrev[pos]
            if prev < 0 { break }
            segments.append(String(chars[prev..<pos]))
            pos = prev
        }

        return segments.reversed()
    }

    public static func split(_ fullName: String, customPath: String? = nil) -> [String] {
        let trimmed = fullName.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return [] }

        let tokens = trimmed.components(separatedBy: .whitespacesAndNewlines).filter { !$0.isEmpty }
        if tokens.count > 1 {
            return tokens
        }

        if LookupIndices.isArabic(trimmed) {
            if LookupIndices.lookup(trimmed, customPath: customPath) != nil {
                return [trimmed]
            }
            return dpSegment(trimmed, customPath: customPath)
        }

        return [trimmed]
    }
}

File: csharp/EgyNames/Splitter.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace EgyptianNames
{
    public static class Splitter
    {
        private const double BaseSegmentCost = 1.0;
        private const double UnknownPenalty = 8.0;
        private const double LengthBonusPerChar = -0.05;

        private static readonly Dictionary<FrequencyClass, double> FreqBonus = new Dictionary<FrequencyClass, double>
        {
            { FrequencyClass.Common, -0.6 },
            { FrequencyClass.Normal, -0.2 },
            { FrequencyClass.Rare, 0.0 }
        };

        private static List<string> DpSegment(string text, string? dataPath = null)
        {
            var arIndex = LookupIndices.GetArForms(dataPath);
            var arNorm = LookupIndices.GetArNormForms(dataPath);

            int n = text.Length;
            var dpCost = new double[n + 1];
            var dpPrev = new int[n + 1];
            var dpKnown = new bool[n + 1];

            for (int i = 0; i <= n; i++)
            {
                dpCost[i] = double.PositiveInfinity;
                dpPrev[i] = -1;
                dpKnown[i] = false;
            }

            dpCost[0] = 0.0;
            dpPrev[0] = 0;
            dpKnown[0] = true;

            for (int i = 1; i <= n; i++)
            {
                int startJ = i > 30 ? i - 30 : 0;
                for (int j = startJ; j < i; j++)
                {
                    if (double.IsPositiveInfinity(dpCost[j])) continue;

                    string substr = text.Substring(j, i - j);
                    if (substr.Length < 2 && j > 0) continue;

                    if (!arIndex.TryGetValue(substr, out var entry))
                    {
                        arNorm.TryGetValue(LookupIndices.NormalizeAr(substr), out entry);
                    }

                    if (entry != null)
                    {
                        double bonus = FreqBonus.TryGetValue(entry.Frequency, out var fb) ? fb : 0.0;
                        double cost = dpCost[j] + BaseSegmentCost + bonus + LengthBonusPerChar * substr.Length;
                        if (cost < dpCost[i])
                        {
                            dpCost[i] = cost;
                            dpPrev[i] = j;
                            dpKnown[i] = true;
                        }
                    }
                    else
                    {
                        double cost = dpCost[j] + UnknownPenalty + substr.Length;
                        if (cost < dpCost[i])
                        {
                            dpCost[i] = cost;
                            dpPrev[i] = j;
                            dpKnown[i] = false;
                        }
                    }
                }
            }

            if (double.IsPositiveInfinity(dpCost[n]))
            {
                return new List<string> { text };
            }

            var segments = new List<string>();
            int pos = n;
            while (pos > 0)
            {
                int prev = dpPrev[pos];
                segments.Add(text.Substring(prev, pos - prev));
                pos = prev;
            }

            segments.Reverse();
            return segments;
        }

        public static List<string> Split(string fullName, string? dataPath = null)
        {
            if (string.IsNullOrWhiteSpace(fullName)) return new List<string>();

            string text = fullName.Trim();

            // If spaces exist, use simple whitespace splitting, but keep a
            // two-word compound lemma (e.g. kunya "Abu X") as one part instead
            // of two meaningless fragments.
            if (text.Contains(" "))
            {
                var raw = text.Split(new[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
                var outList = new List<string>();
                int i = 0;
                int n = raw.Length;
                while (i < n)
                {
                    if (i < n - 1)
                    {
                        var pair = $"{raw[i]} {raw[i + 1]}";
                        if (LookupIndices.LookupAr(pair, dataPath) != null || LookupIndices.LookupAr($"{raw[i]}{raw[i + 1]}", dataPath) != null)
                        {
                            outList.Add(pair);
                            i += 2;
                            continue;
                        }
                    }
                    outList.Add(raw[i]);
                    i++;
                }
                return outList;
            }

            if (LookupIndices.IsArabic(text))
            {
                var entry = LookupIndices.Lookup(text, dataPath);
                if (entry != null)
                {
                    return new List<string> { text };
                }
                return DpSegment(text, dataPath);
            }

            return new List<string> { text };
        }
    }
}
